### PR TITLE
core,netty,okhttp: detect proxy via ProxySelector

### PR DESCRIPTION
**Pull request by zpencer**
For DNS names: use the system wide ProxySelector to determine the proxy. GRPC_PROXY_EXP is treated as an override, but will be removed in the future.

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/Proxies.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  /**
+   * Given a target address, returns which proxy address should be used. If no proxy should be
+   * used, then return null.
+   */
+  public static InetSocketAddress proxyFor(SocketAddress targetServerAddress) {
+    return proxyFor(
+        targetServerAddress,
+        ProxySelector.getDefault(),
+        System.getenv(GRPC_PROXY_ENV_VAR)
+    );
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/Proxies.java]: We mostly put the parenthesis on the same line as the last argument. Since you can't have a comma after the getenv, the newline doesn't really seem to help anything.

File: core/src/main/java/io/grpc/internal/Proxies.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+  /**
+   * Helper method to achieve stateless unit tests.
+   */
+  @VisibleForTesting
+  static InetSocketAddress proxyFor(
+      SocketAddress targetServerAddress,
+      ProxySelector proxySelector,
+      @Nullable String proxyEnvStr
+  ) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/Proxies.java]: Ditto.

File: core/src/main/java/io/grpc/internal/Proxies.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+  private static InetSocketAddress detectProxy(
+      ProxySelector proxySelector,
+      SocketAddress targetServerAddress
+  ) {
+    if (!(targetServerAddress instanceof InetSocketAddress)) {
+      return null;
+    }
+
+    String hostName = ((InetSocketAddress) targetServerAddress).getHostName();
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/Proxies.java]: If the socket address was created with an IP address this can do I/O. If you only expect this class to work with `DnsNameResolver`, you could `assert address.isUnresolved()`. But this problem has been causing some havoc. There is getHostString(), but it's not available in Java 6.

File: core/src/main/java/io/grpc/internal/Proxies.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    URI uri;
+    try {
+      uri = new URI(String.format(URI_FORMAT, hostName));
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/Proxies.java]: This string.format does not work for IPv6 addresses. Use `GrpcUtil.authorityFromHostAndPort()`

Comment (ejona86) [inline core/src/main/java/io/grpc/internal/Proxies.java]: This should include the port. The port may change the proxy decision.

File: core/src/main/java/io/grpc/internal/Proxies.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    } catch (final URISyntaxException e) {
+      log.warning(String.format(
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/Proxies.java]: Always try to use the logger's formatter, as it is good habit to reduce chances of doing something like this with INFO and lower which will cause useless computation:

```java
log.log(Level.WARNING, "blah: {0}", hostName);
```

File: netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java [hunk 6]
```diff
@@ -365,20 +366,16 @@ void overrideAuthorityChecker(@Nullable OverrideAuthorityChecker authorityChecke
...
+    SocketAddress proxy = Proxies.proxyFor(targetServerAddress);
     if (proxy != null) {
-      String[] parts = proxy.split(":", 2);
-      int port = 80;
-      if (parts.length > 1) {
-        port = Integer.parseInt(parts[1]);
-      }
-      InetSocketAddress proxyAddress = new InetSocketAddress(parts[0], port);
-      negotiator = ProtocolNegotiators.httpProxy(proxyAddress, null, null, negotiator);
+      negotiator = ProtocolNegotiators.httpProxy(proxy, null, null, negotiator);
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java]: ProxySelector assumes you will be using Authenticator as well. We can break setups if we use ProxySelector without using Authenticator.

File: netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java [hunk 7]
```diff
@@ -365,20 +366,16 @@ void overrideAuthorityChecker(@Nullable OverrideAuthorityChecker authorityChecke
...
+      SocketAddress targetServerAddress,
       String authority,
       NegotiationType negotiationType,
-      SslContext sslContext) {
+      SslContext sslContext
+  ) {
     ProtocolNegotiator negotiator =
         createProtocolNegotiatorByType(authority, negotiationType, sslContext);
-    String proxy = System.getenv("GRPC_PROXY_EXP");
+    SocketAddress proxy = Proxies.proxyFor(targetServerAddress);
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java]: This makes me worried for tests, since this uses ProxySelector with its global auto-configuration.

File: core/src/main/java/io/grpc/internal/Proxies.java (hunk 1 above)
Comment (zpencer) [inline core/src/main/java/io/grpc/internal/Proxies.java]: done

File: core/src/main/java/io/grpc/internal/Proxies.java (hunk 5 above)
Comment (zpencer) [inline core/src/main/java/io/grpc/internal/Proxies.java]: Ah didn't realize this was the way to do it here, it's a bit different from log4j

File: core/src/main/java/io/grpc/internal/Proxies.java (hunk 2 above)
Comment (zpencer) [inline core/src/main/java/io/grpc/internal/Proxies.java]: done, and audited rest of changes

File: core/src/main/java/io/grpc/internal/Proxies.java (hunk 4 above)
Comment (zpencer) [inline core/src/main/java/io/grpc/internal/Proxies.java]: done

Comment (zpencer) [inline core/src/main/java/io/grpc/internal/Proxies.java]: done

File: netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java (hunk 7 above)
Comment (zpencer) [inline netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java]: Updated diff to use dependency injection to avoid static calls

File: netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java (hunk 6 above)
Comment (zpencer) [inline netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java]: Updated diff to include changes for Authenticator, which was originally intended separate diff.

File: core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+    String hostPort =
+        GrpcUtil.authorityFromHostAndPort(targetAddr.getHostName(), targetAddr.getPort());
+    URI uri;
+    try {
+      uri = new URI(String.format(URI_FORMAT, hostPort));
+    } catch (final URISyntaxException e) {
+      log.log(
+          Level.WARNING,
+          "Failed to construct URI for proxy lookup, proceeding without proxy: {0}",
+          hostPort);
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java]: It will be useful to log the exception.

File: core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+    if (auth == null) {
+      return new ProxyParameters(proxyAddr, null, null);
+    }
+
+    try {
+      //todo(spencerfang): users of ProxyParameters should clear the password when done
+      return new ProxyParameters(proxyAddr, auth.getUserName(), new String(auth.getPassword()));
+    } finally {
+      Arrays.fill(auth.getPassword(), '0');
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java]: Why is this necessary?

File: core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * A utility class that detects proxies using {@link ProxySelector} and detects authentication
+ * credentials using {@link Authenticator}.
+ *
+ */
+public class ProxyDetectorImpl implements ProxyDetector {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java]: This can be package-private.

File: core/src/main/java/io/grpc/internal/ProxyDetector.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+package io.grpc.internal;
+
+import com.google.common.base.Objects;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import javax.annotation.Nullable;
+
+/**
+ * A utility class to detect which proxy, if any, should be used for a given
+ * {@link java.net.InetSocketAddress}.
```
Comment (zpencer) [inline core/src/main/java/io/grpc/internal/ProxyDetector.java]: Done

File: core/src/main/java/io/grpc/internal/ProxyDetector.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+  ProxyDetector NOOP_INSTANCE = new ProxyDetector() {
+    @Override
+    public ProxyParameters proxyFor(SocketAddress targetServerAddress) {
+      return null;
+    }
+  };
+
+  /**
+   * Given a target address, returns which proxy address should be used. If no proxy should be
+   * used, then return value will be an absent value.
```
Comment (zpencer) [inline core/src/main/java/io/grpc/internal/ProxyDetector.java]: Done

File: core/src/main/java/io/grpc/internal/ProxyDetector.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+   */
+  ProxyDetectorImpl.ProxyParameters proxyFor(SocketAddress targetServerAddress);
```
Comment (zpencer) [inline core/src/main/java/io/grpc/internal/ProxyDetector.java]: Done

File: core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+    PasswordAuthentication requestPasswordAuthentication(
+        String host,
+        InetAddress addr,
+        int port,
+        String protocol,
+        String prompt,
+        String scheme);
+  }
+
+  static class AuthenticationProviderImpl implements AuthenticationProvider {
```
Comment (zpencer) [inline core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java]: Done, there's no reason why this can't be anonymous

File: core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java (hunk 9 above)
Comment (zpencer) [inline core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java]: The javadoc for PasswordAuthentication recommends clearing the password after it is used. Even though gRPC doesn't treat passwords in a special way today, I think having one fewer place where passwords linger around is better.

**Comment by zpencer** [comment]
There's some more changes coming to this branch to avoid using ProxyDetector twice, and instead plumb the ProxyParameters into the ClientTransportFactory via Attributes.

File: core/src/main/java/io/grpc/internal/DnsNameResolver.java [hunk 15]
```diff
@@ -160,9 +164,10 @@ public void run() {
           resolving = true;
         }
         try {
-          if (System.getenv("GRPC_PROXY_EXP") != null) {
-            EquivalentAddressGroup server =
-                new EquivalentAddressGroup(InetSocketAddress.createUnresolved(host, port));
+          InetSocketAddress destination = InetSocketAddress.createUnresolved(host, port);
+          ProxyParameters proxyParams = proxyDetector.proxyFor(destination);
+          if (proxyParams != null) {
+            EquivalentAddressGroup server = new EquivalentAddressGroup(destination);
```
Comment (zpencer) [inline core/src/main/java/io/grpc/internal/DnsNameResolver.java]: #3030 will change the flow here

**Comment by zpencer** [comment]
PTAL: This diff now addresses case 1 and 2, while making it easy to implement 3 in the future. 

`ClientTransportFactory #newClientTransport` will now require a `ProxyParameters` argument. Implementations are responsible for correctly connecting to `serverAddress` via the specified proxy.

For case 1:
- This happens when the DNS name is associated with a HTTP proxy. `DnsNameResolver` will see that a proxy should be used, and there will be 1 EAG containing a single entry, the unresolved destination address. This is a bit different from c-core, because in c-core the resolved address is the proxy's address.
- In general, the name resolver implementation is responsible for figuring out whether a proxy is used. We can't do the check in MangedChannelImpl because the `target` string is opaque, and it is up to the NameResolver implementations to turn the string into host(s).
- Later when it is time for the `InternalSubchannel` to connect, the class will consult with the `ProxyDetector` to determine the proxy settings to pass into `newClientTransport`.

For case 2:
- This happens when the DNS name is not associated with any proxies, and `DnsNameResolver` is able to resolve the name to IPs and ports.
- The `ProxyDetector` must have the IP:PORT resolution results associated with the correct proxy values, so that `InternalSubchannel` can look up the proxy settings to pass into `newClientTransport`.

For case 3:
- Connecting to the LB: As long as `ProxyDetector` has a proxy associated with the LB's hostname, the name resolver will leave the name unresolved. `GrpcLoadBalancer` uses `InternalSubchannel` to create out of band channels with the LB, which means the proxy settings will be correctly autodetected.
- Connecting to the backends: As long as each backend server returned by the LB each have a proxy associated with it, `InternalSubchannel` will detect the proxy and pass it into `newClientTransport`.

**Comment by zpencer** [comment]
ping @ejona86

File: core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+ */
+class ProxyDetectorImpl implements ProxyDetector {
+  private static final Logger log = Logger.getLogger(ProxyDetectorImpl.class.getName());
+  private static final AuthenticationProvider DEFAULT_AUTHENTICATOR = new AuthenticationProvider() {
+    @Override
+    public PasswordAuthentication requestPasswordAuthentication(
+        String host, InetAddress addr, int port, String protocol, String prompt, String scheme) {
+      // TODO(spencerfang): consider using java.security.AccessController here
+      return Authenticator.requestPasswordAuthentication(
+          host, addr, port, protocol, prompt, scheme);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java]: You should specify RequestorType.PROXY

File: core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+      log.warning("More than 1 proxy detected, gRPC will select the first one");
+    }
+    Proxy proxy = proxies.get(0);
+
+    if (proxy.type() == Proxy.Type.DIRECT) {
+      return null;
+    }
+    InetSocketAddress proxyAddr = (InetSocketAddress) proxy.address();
+    PasswordAuthentication auth = authenticationProvider.requestPasswordAuthentication(
+        proxyAddr.getHostName(),
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java]: This may do a DNS request. I wonder if this is an appropriate time to do reflection and check whether `getHostString()` is available... Java 6 does `getHostName()` here, but Java 7 swaps to `getHostString()`.

File: core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+        proxyAddr.getAddress(),
+        proxyAddr.getPort(),
+        PROXY_SCHEME,
+        "proxy authentication required by gRPC",
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java]: This is supposed to be the realm returned by the server. We won't have it since we are avoiding the full authentication handshake. We maybe should use `""`, but at the very least it deserves a comment.
